From: Jeff Mahoney <jeffm@suse.com>
Subject: [PATCH 16/15] ocfs2: include disk heartbeat in ocfs2_nodemanager to avoid userspace changes

 This patch removes disk heartbeat's modularity which makes it the default.

 Without this patch, userspace changes are required.

 This patch is not intended for permanent application, just to make it easier
 for users not interested in testing the userspace clustering implementation
 to use ocfs2.

 In order to switch to user clustering, use "o2cb offline" to shut down the
 cluster, echo "user" > /sys/o2cb/heartbeat_mode, and "o2cb online" to
 start it back up again.

 fs/Kconfig                        |   12 ------------
 fs/ocfs2/cluster/Makefile         |    3 +--
 fs/ocfs2/cluster/disk_heartbeat.c |    8 ++------
 fs/ocfs2/cluster/disk_heartbeat.h |    2 ++
 fs/ocfs2/cluster/nodemanager.c    |    7 +++++++
 5 files changed, 12 insertions(+), 20 deletions(-)

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

--- a/fs/Kconfig	2007-08-27 14:01:22.000000000 -0400
+++ b/fs/Kconfig	2007-08-27 14:03:22.000000000 -0400
@@ -473,18 +473,6 @@ config OCFS2_DEBUG_MASKLOG
 	  This option will enlarge your kernel, but it allows debugging of
 	  ocfs2 filesystem issues.
 
-config OCFS2_FS_O2CB
-	tristate "O2CB Kernelspace Clustering"
-	depends on OCFS2_FS
-	default y
-	help
-	  OCFS2 includes a simple kernelspace clustering package that requires
-	  a very small userspace complement to configure it. However, it is
-	  not as flexible as an external cluster manager, and is only
-	  limited to maintaining a cluster for OCFS2 file systems.
-	  If you are not using an external clustering package, choose
-	  this option.
-
 config OCFS2_FS_USERSPACE_CLUSTER
 	tristate "Userspace Clustering"
 	depends on OCFS2_FS
--- a/fs/ocfs2/cluster/Makefile	2007-08-27 14:01:22.000000000 -0400
+++ b/fs/ocfs2/cluster/Makefile	2007-08-27 14:01:22.000000000 -0400
@@ -3,7 +3,6 @@ obj-$(CONFIG_OCFS2_FS_O2CB) += ocfs2_dis
 obj-$(CONFIG_OCFS2_FS_USERSPACE_CLUSTER) += ocfs2_user_heartbeat.o
 
 ocfs2_nodemanager-objs := heartbeat.o masklog.o sys.o nodemanager.o \
-	tcp.o ver.o
+	tcp.o ver.o disk_heartbeat.o quorum.o
 
-ocfs2_disk_heartbeat-objs := disk_heartbeat.o quorum.o
 ocfs2_user_heartbeat-objs := user_heartbeat.o
--- a/fs/ocfs2/cluster/disk_heartbeat.c	2007-08-27 14:01:22.000000000 -0400
+++ b/fs/ocfs2/cluster/disk_heartbeat.c	2007-08-27 14:01:22.000000000 -0400
@@ -1505,7 +1505,7 @@ void o2hb_stop_all_regions(void)
 	spin_unlock(&o2hb_live_lock);
 }
 
-static int __init o2hb_disk_heartbeat_init(void)
+int o2hb_disk_heartbeat_init(void)
 {
 	int i;
 
@@ -1516,14 +1516,10 @@ static int __init o2hb_disk_heartbeat_in
 	return o2hb_register_heartbeat_group(&disk_heartbeat_group);
 }
 
-static void __exit o2hb_disk_heartbeat_exit(void)
+void o2hb_disk_heartbeat_exit(void)
 {
 	o2hb_unregister_heartbeat_group(&disk_heartbeat_group);
 }
 
 MODULE_AUTHOR("Oracle");
 MODULE_LICENSE("GPL");
-
-module_init(o2hb_disk_heartbeat_init);
-module_exit(o2hb_disk_heartbeat_exit);
-
--- a/fs/ocfs2/cluster/disk_heartbeat.h	2007-08-27 14:01:22.000000000 -0400
+++ b/fs/ocfs2/cluster/disk_heartbeat.h	2007-08-27 14:01:22.000000000 -0400
@@ -36,5 +36,7 @@ extern unsigned int o2hb_dead_threshold;
 #define O2HB_MIN_DEAD_THRESHOLD	  2
 #define O2HB_MAX_WRITE_TIMEOUT_MS (O2HB_REGION_TIMEOUT_MS * (o2hb_dead_threshold - 1))
 void o2hb_stop_all_regions(void);
+int o2hb_disk_heartbeat_init(void);
+void o2hb_disk_heartbeat_exit(void);
 
 #endif /* O2CLUSTER_DISK_HEARTBEAT_H */
--- a/fs/ocfs2/cluster/nodemanager.c	2007-08-27 14:01:22.000000000 -0400
+++ b/fs/ocfs2/cluster/nodemanager.c	2007-08-27 14:01:22.000000000 -0400
@@ -29,6 +29,7 @@
 #include "nodemanager.h"
 #include "heartbeat.h"
 #include "quorum.h"
+#include "disk_heartbeat.h"
 #include "masklog.h"
 #include "sys.h"
 #include "ver.h"
@@ -906,6 +907,7 @@ static void __exit exit_o2nm(void)
 
 	/* XXX sync with hb callbacks and shut down hb? */
 	o2net_unregister_hb_callbacks();
+	o2hb_disk_heartbeat_exit();
 	configfs_unregister_subsystem(&o2nm_cluster_group.cs_subsys);
 	o2cb_sys_shutdown();
 	o2net_exit();
@@ -918,6 +920,10 @@ static int __init init_o2nm(void)
 	cluster_print_version();
 
 	o2hb_init();
+	ret = o2hb_disk_heartbeat_init();
+	if (ret)
+		goto out;
+
 	o2net_init();
 
 	ocfs2_table_header = register_sysctl_table(ocfs2_root_table);
@@ -950,6 +956,7 @@ out_sysctl:
 	unregister_sysctl_table(ocfs2_table_header);
 out_o2net:
 	o2net_exit();
+	o2hb_disk_heartbeat_exit();
 out:
 	return ret;
 }
